core/trino-main/src/main/java/io/trino/operator/output/columnar/ColumnarPagePartitioner.java

```diff
@@ -0,0 +1,538 @@
+        }
+        return partitionPositions;
+    }
+
+    private static int initialPartitionSize(int averagePositionsPerPartition)
+    {
+        // 1.1 coefficient compensates for the not perfect hash distribution.
+        // 32 compensates for the case when averagePositionsPerPartition is small,
+        // and we would see more variance in the hash distribution.
+        return (int) (averagePositionsPerPartition * 1.1) + 32;
+    }
+
+    private boolean onlyRleBlocks(Page page)
+    {
+        for (int i = 0; i < page.getChannelCount(); i++) {
+            if (!(page.getBlock(i) instanceof RunLengthEncodedBlock)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void partitionBySingleRleValue(Page page, int position, Page partitionFunctionArgs, IntArrayList[] partitionPositions)
+    {
+        // copy all positions because all hash function args are the same for every position
+        if (nullChannel != -1 && page.getBlock(nullChannel).isNull(0)) {
+            verify(page.getBlock(nullChannel) instanceof RunLengthEncodedBlock, "null channel is not RunLengthEncodedBlock", page.getBlock(nullChannel));
+            // all positions are null
+            int[] allPositions = integersInRange(position, page.getPositionCount());
+            for (IntList partitionPosition : partitionPositions) {
+                partitionPosition.addElements(position, allPositions);
+            }
+        }
+        else {
+            // extract rle page to prevent JIT profile pollution
+            Page rlePage = extractRlePage(partitionFunctionArgs);
+
+            int partition = partitionFunction.getPartition(rlePage, 0);
+            IntArrayList positions = partitionPositions[partition];
+            for (int i = position; i < page.getPositionCount(); i++) {
+                positions.add(i);
+            }
+        }
+    }
+
+    private Page extractRlePage(Page page)
+    {
+        Block[] valueBlocks = new Block[page.getChannelCount()];
+        for (int channel = 0; channel < valueBlocks.length; ++channel) {
+            valueBlocks[channel] = ((RunLengthEncodedBlock) page.getBlock(channel)).getValue();
+        }
+        return new Page(valueBlocks);
+    }
+
+    private int[] integersInRange(int start, int endExclusive)
+    {
+        int[] array = new int[endExclusive - start];
+        int current = start;
+        for (int i = 0; i < array.length; i++) {
+            array[i] = current++;
+        }
+        return array;
+    }
+
+    private boolean isDictionaryProcessingFaster(Block block)
+    {
+        if (!(block instanceof DictionaryBlock)) {
+            return false;
+        }
+        DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+        // if dictionary block positionCount is greater than number of elements in the dictionary
```

Comment:
That is not always true. A block may have a giant dictionary and use only a small subset of values. Please add a note that this is only a heuristic

Reply:
we can't have giant dictionary blocks here as the dictionary size is limited by page size by the if below.

Reply:
nit: in the future we could reuse dictionaries between pages